app/src/main/java/de/westnordost/streetcomplete/quests/smoking/AddSmoking.kt

```diff
@@ -28,6 +28,7 @@ class AddSmoking : OsmFilterQuestType<SmokingAllowed>() {
              )
          )
          and takeaway != only
+         and !(indoor_seating = no and outdoor_seating = no)
```

Comment:
See [#3821](https://togithub.com/streetcomplete/StreetComplete/pull/3821):

> I wanted to do
> `and !sidewalk and !sidewalk:both and !(sidewalk:left and sidewalk:right)`
> but the NOT operator doesn't seem to work with brackets (it complained I'd closed too many brackets).

Reply:
You should be able to rewrite it with [De Morgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) though:

```suggestion
         and (indoor_seating != no or outdoor_seating != no)
```

Reply:
Maybe, but I'm uncomfortable because I don't know how that parsing code handles null (undef) values, that's why I written it in what seems to me the "safe" way...

Also I find it much easier to read (given the fact that tags are not boolean `yes` / `no` values, but can have many values which differ from `no`, or no value at all).

Reply:
But interesting issue #3821, has it been **known** that `!` does not work with brackets, or was that just a _hypothesis_? Because it didn't complain in my case. I plan to do some checking how it actually behaves tonight to make sure.

Reply:
When you work it out, it would be good to add some tests for that (if they don't exist already), to ensure it continues to behave in the same way in case this or something else is making use of that behaviour.

Reply:
`!` doesn't work with parenthesis, parsing for this is simply not implemented. It was known to me at least :-)
I once started to implement it because I thought it would be "cleaner" to also support parsing this despite it not being used that far but I stopped because it turned out to be more complex than I anticipated.

`indoor_seating = no` evaluates to true if `indoor_seating` is either not set or not `no`.

Reply:
thanks, this should be incorporated now... Now to testing if it actually works like it should